### PR TITLE
Add verify-knife script and enforce LF line endings

### PR DESCRIPTION
**Pull request by BTernaryTau**
Since there seems to be support for standardizing on LF line endings in working directories, I'm creating this as an alternative to #4209.

Once this PR or #4209 has been merged, the other one can simply be closed.

### CODE REVIEW

**Review by jkingdon — APPROVED** [review]

**Comment by jkingdon** [comment]
As I said on the other pull request, if people who are more directly affected by the `text eol=lf` part of this than me have stronger opinions than mine, I'll defer to them.

File: scripts/verify-knife
Comment (benjub) [inline scripts/verify-knife:24]:
```suggestion
  echo "The file '${discouraged_file}' needs to be regenerated."
```

**Review by benjub — COMMENTED** [review]
Both the yml and bash look good to me, and I prefer that version over the other MR, but not a bash expert here.

File: scripts/verify-knife
Comment (wlammen) [inline scripts/verify-knife:20]: Why is this command guaranteed to always succeed, and has no return value check?  Even if the current implementation never fails, a future version might do.
Or use && to start the diff only if metamath-knife succeeds, for example
   metamath-knife ..params.. && diff ..params..

File: scripts/verify-knife
Comment (jkingdon) [reply scripts/verify-knife:24]: Heh. This is a dialect variation of different varieties of English so I'm OK with either one.

File: scripts/verify-knife
Comment (wlammen) [inline scripts/verify-knife:16]: It is slightly better to use -ne instead of !=, because != does a string comparison, which needs a prior conversion from integer to string, while -ne does a numeric comparison.

File: scripts/verify-knife
Comment (wlammen) [inline scripts/verify-knife:8]: Is it possible to start this script outside the metamath folder?  What if one of the submitted files does not exist, say because of a spelling mistake?

File: scripts/verify-knife
Comment (benjub) [reply scripts/verify-knife:24]: It's already hard enough to learn a foreign language, not to have to also learn its dialectal variations.  Please, think of us non-native speakers !

File: scripts/verify-knife
Comment (BTernaryTau) [reply scripts/verify-knife:8]: Yes, many (most?) scripts assume by default that they are being run from the metamath folder, but that's not a requirement.

If the source file does not exist:
```
error: I/O error: The system cannot find the file specified. (os error 2)
--> missing-source:1:1
 |
 |
1 diagnostics issued.
```

If the discouraged file does not exist:
```
0 diagnostics issued.
diff: missing-discouraged: No such file or directory
The file 'missing-discouraged' needs to be regenerated.
```

Both cases exit with status 1.

File: scripts/verify-knife
Comment (BTernaryTau) [reply scripts/verify-knife]: Done.

File: scripts/verify-knife
Comment (BTernaryTau) [reply scripts/verify-knife]: It's not, the intent was that failures are caught by diff not finding discouraged.new. I updated it to use &&.

File: scripts/verify-knife
Comment (BTernaryTau) [reply scripts/verify-knife]: I just went with the shorter one, but if it's less readable then I agree it's worth updating.

File: scripts/verify-knife
Comment (wlammen) [reply scripts/verify-knife:8]: The reason I mentioned this, is that this handling is not obvious.  So it belongs to the introducing description, like
# If one of the files does not exist, the script terminates with exit code 1.

The assumptions the script makes on start up are called pre-conditions,  and guarantees about the state on termination are called post-conditions.  The post-condition here is that the script checks the integrity of the .mm file in some (which?)  respects, and if that test succeeds and the discouraged file matches (in what way? Is its format fixed?), then, and only then, the script terminates with exit code 0.

Both conditions are written down in the introducing comment.

The following is a bit nit-picking, but maybe you are still on the learning side.  The clean-up code rm discouraged.new is duplicated, and that is not necessary:

metamath-knife && diff
retval="$?"
rm discouraged.new
exit $retval

Comment (BTernaryTau) [reply scripts/verify-knife:17]: I added a description to the script file. I didn't change the cleanup logic because I want the script to match metamath-knife's behavior when a file is missing (exit with status 1) rather than diff's behavior (exit with status 2).

Comment (wlammen) [reply scripts/verify-knife:17]: This idea is also worth being added as a comment.  It is hard to guess.

You can still use the tactics I depicted to have the clean-up in just one place.  The early out on error strategy of your code is less maintainable in more complex contexts, without something like an atexit at hand.  See  this post https://stackoverflow.com/questions/2129923/how-to-run-a-command-before-a-bash-script-exits for a more general approach.

All my remarks given here (and the days before) are neither approvals nor disapprovals.  So you can (could have) follow or ignore them.  I just tried to be convincing.  The code is now in a good enough shape that it can be committed from my point of view.  Adjustments to follow some advanced code style can be made afterwards.  Folks should now decide which of the two pull requests is preferred.

Comment (benjub) [reply scripts/verify-knife:17]: AFAIK, it's up to you @BTernaryTau: can we merge, or do you plan to make other changes to this MR, following @wlammen's proposals ?

(As for me, this version is better than #4209)

Comment (BTernaryTau) [reply scripts/verify-knife:17]: I'm good to merge it as is at this point.

Comment (jkingdon) [reply scripts/verify-knife:17]: I'll take that as my cue. Done.

**Pull request merged**
